### PR TITLE
feat: workflows can be created without associated dependent resource

### PR DESCRIPTION
**Pull request by metacosm**
This allows for the workflow graph to be resolved at build time from a 
factory, the dependent resources being configured as needed at runtime
via the resolve mechanism.

### CODE REVIEW

**Review by csviri — APPROVED** [review]
Apart from little naming LGTM

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/DependentResource.java [hunk 1]
```diff
@@ -64,4 +64,8 @@ default Optional<R> getSecondaryResource(P primary, Context<P> context) {
   static String defaultNameFor(Class<? extends DependentResource> dependentResourceClass) {
     return dependentResourceClass.getName();
   }
+
+  static boolean canDeleteIfAble(DependentResource<?, ?> dependentResource) {
```
[the remark this answers is not in the document]
Comment (metacosm) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/DependentResource.java:68]: What would you name it?

[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/DependentResource.java:68]: `deletable`?

[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/DependentResource.java:68]: Also problem here is that GarbageCollected is only applicatble for KubernetesDependentResource, so it is little smelly that it's on this layer.

**Review by csviri — CHANGES_REQUESTED** [review]

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/DependentResource.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (metacosm) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/DependentResource.java]: Yeah, wasn't too sure where to put it but the problem is that the code in this method was spread at different spots that didn't have anything to do with `KubernetesDependentResource` either so it's less smelly now than it was.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/DependentResource.java [hunk 2]
```diff
@@ -64,4 +64,8 @@ default Optional<R> getSecondaryResource(P primary, Context<P> context) {
...
+    return dependentResource instanceof Deleter && !(dependentResource instanceof GarbageCollected);
```
[the remark this answers is not in the document]
Comment (metacosm) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/DependentResource.java:69]: This isn't the case in the tests, though…

[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/DependentResource.java:69]: Hmm where exactly?

[the remark this answers is not in the document]
Comment (metacosm) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/DependentResource.java]: See https://github.com/java-operator-sdk/java-operator-sdk/pull/1632/commits/0a1a381d48dc5d3a426d7a8ad7c5b468be3a2061

**Review by csviri — APPROVED** [review]
LGTM

**Pull request merged**